lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -0,0 +1,229 @@
+#
+# This file represents the output of the sysctl lib, which can combine multiple
+# configurations into a single file like.
+"""
+
+
+class Error(Exception):
+    """Base class of most errors raised by this library."""
+
+    def __repr__(self):
+        """Represent the Error."""
+        return "<{}.{} {}>".format(type(self).__module__, type(self).__name__, self.args)
+
+    @property
+    def name(self):
+        """Return a string representation of the model plus class."""
+        return "<{}.{}>".format(type(self).__module__, type(self).__name__)
+
+    @property
+    def message(self):
+        """Return the message passed as an argument."""
+        return self.args[0]
+
+
+class SysctlError(Error):
+    """Raised when there's an error running sysctl command."""
+
+
+class SysctlPermissionError(Error):
+    """Raised when there's an error applying values in sysctl."""
+
+
+class ValidationError(Error):
+    """Exception representing value validation error."""
+
+
+class SysctlConfig(Mapping[str, int]):
+    """Represents the state of the config that a charm wants to enforce."""
+
+    def __init__(self, name: str = "") -> None:
+        self.name = name
+
+    def __contains__(self, key: str) -> bool:
+        """Check if key is in config."""
+        return key in self._data
+
+    def __len__(self):
+        """Get size of config."""
+        return len(self._data)
+
+    def __iter__(self):
+        """Iterate over config."""
+        return iter(self._data)
+
+    def __getitem__(self, key: str) -> int:
+        """Get value for key form config."""
+        return self._data[key]
+
+    @property
+    def _data(self) -> Dict[str, int]:
+        """Return applied internal config."""
+        if not self.merged_config_exists:
+            return {}
+
+        with open(SYSCTL_FILENAME, "r") as f:
+            return {param.strip(): int(value.strip())
+                    for line in f.read().splitlines()
+                    if line and not line.startswith('#')
+                    for param, value in [line.split('=')]}
+
+    @property
+    def name(self) -> str:
+        """Name used to create the lib file."""
+        return self._name
+
+    @name.setter
+    def name(self, value: str):
+        if value == "":
+            self._name, *_ = os.getenv("JUJU_UNIT_NAME", "unknown").split("/")
+        else:
+            self._name = value
+
+    @property
+    def charm_filepath(self) -> str:
+        """Name for resulting charm config file."""
+        return f"{SYSCTL_DIRECTORY}/90-juju-{self.name}"
+
+    @property
+    def charm_config_exists(self) -> bool:
+        """Return whether the charm config file exists."""
+        return os.path.exists(self.charm_filepath)
+
+    @property
+    def merged_config_exists(self) -> bool:
+        """Return whether a merged config file exists."""
+        return os.path.exists(SYSCTL_FILENAME)
+
+    def update(self, config: Dict[str, dict]) -> None:
```

Comment:
**thought** I can see some edge cases happening here (similar to the points of Robert below). I'm thinking that I believe it would make the code a bit more robust to refine slightly the business logic, to do things in memory and try to leave the status of the system as clean as possible

1. When we initialize (init) the object we read (if the current configuration exists, file `SYSCTL_FILENAME`) the status of the system. I believe it makes sense to have a status property for this object that reflects current configuration based on the sysctl_filename. 
2. On update, we take the new config (in memory), we read all other configuration files that are present (sysctl_directory/90*), we merge and validate the config IN MEMORY (here we should read the sysctl_filename as if we are changing a value for my charm, that would of course conflicts). Here if there is a ValidationError exception (when merging in memory), we just raise such that charm authors can move to blocked status, but most importantly we did not modify the state on the system yet (we did not modify the sysctl filename or created a new charm_filepath file, possibly impacting other charms as well) 
3. If everything was fine  (so no conflicts), we write the sysctl_file (with proper header, stating the names of the files that contributed to it)
4. we try to apply it. If it breaks, we roll back, using the state of the system (e.g. _data), we don't overwrite the `charm_filepath` file (otherwise this may again break existing charms) and we raise SysctlError. If it works fine, we update the internal _data field and we write the `charm_filepath` file (such that any other merge, from others charm will still work).  

(There is still some race conditions that may be happening between 3 and 4, but this should really be rather quick if "applying" is fast, so it should be rather remote that this may create problems)

**niptick** Maybe, one could make just write a `merge` function (but that raises an exception if the key already exists with a different value) based on dicts, e.g. `def _merge(*configs: dict[str, int]) -> dict[str, int]`. The input `configs*` could either come from files or the input of the update.

Reply:
Validation is already done in memory, see `_validate()` method.
For applying the config, it has to be done after validation and before merging, as there are keys that will only fail when calling `sysctl <key>=<value>`. In essence, `_merge()` will just write the final file that has been applied, it won't apply anything, as the `apply` is done on the previous step.

Reply:
yes, but if I understand the code correctly: _validate() keeps on reading the file (so it is also kinda slow) and - because of this - you are a bit more prone to racing conditions. It just feels more effective to (1) read in memory (2) validate (without keep on reading) (3) applying (4) consolidate/rollback

Reply:
Updated the lib with _load_data() calls so it's only done once.